[build][c10] set up c10 scaffolding. Move macros proper first.

Yangqing opened the pull request.

Yangqing commented on CMakeLists.txt:
```diff
@@ -5,6 +5,12 @@ cmake_minimum_required(VERSION 3.5 FATAL_ERROR)
+if (NOT MSVC)
+  set(CMAKE_C_STANDARD 11)
+endif()
+set(CMAKE_VERBOSE_MAKEFILE ON)
```
(note: this will be removed when the pr is ready to be merged)

ezyang commented on aten/src/ATen/core/Macros.h:
```diff
@@ -3,41 +3,7 @@
-#else // defined(CAFFE2_BUILD_MAIN_LIBS) || defined(ATen_cpu_EXPORTS) || defined(caffe2_EXPORTS)
-#define AT_CORE_API AT_CORE_IMPORT
-#endif // defined(CAFFE2_BUILD_MAIN_LIBS) || defined(ATen_cpu_EXPORTS) || defined(caffe2_EXPORTS)
+#include "c10/macros/Macros.h"
```
What's the rule for subfolders under c10? My preference (evidenced by `ATen/core/` structure) is that most headers live in the top level.

Yangqing replied:
At some point we should just include a unified c10/c10.h that includes all dependent headers, and have c10 manage the internal header relationships - does that make sense? I think eigen uses this as well by including headers like <Eigen/Core> without exposing further details.

Yangqing committed:
set up c10 scaffolding. Move macros proper first.

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/11939

Differential Revision: D10004629

fbshipit-source-id: f9ba2ccfc33cf1e7620bf3b271cf370ae422f863

Yangqing commented:
@pytorchbot retest this please. Seems like a CI system error on mac.

Yangqing commented:
OK, this one should be correct because last run had mac errors due to ci system, this time mac passed, and mkl is erroring out due to timeout, but mkl already passed last time. No code change between the two runs.
